### ISSUE TITLE
Question: Using StringWriter to optimize some hot paths, but running into double serialization

### ISSUE THREAD

**Issue by ElMassimo** [ohler55/oj#542]
#### Config ⚙️ 
Using `Oj.optimize_rails`.

#### Description 📖 
`Oj::StringWriter` is great to avoid creating intermediate `Hash` objects, but I'm running into a problem when doing something similar to:

```ruby
writer = Oj::StringWriter.new
...
render json: { key: writer.to_s }
```

To prevent double serialization to JSON, I attempted to wrap the string in an object:

```ruby
class JsonValue
  def initialize(json)
    @json = json
  end

  def as_json(_options = nil)
    self
  end

  def to_json(*)
    @json
  end
end
``` 

and used it like this:

```ruby
writer = Oj::StringWriter.new
...
render json: { key: JsonValue.new(writer.to_s) }
```

However, it looks like the methods are bypassed by the optimization.

### Question ❔ 
Is there a special class that we can use to signal `Oj` that a value is already serialized and should be dumped as is? (similar to `push_json` in `Oj::StringWriter`)

**Comment by ohler55** [comment]
When using `Oj.optimize_rails` you will be in rails mode. I haven't tested but I depending on how you call Oj, the default options mode will be ignored. I assume you are using rails and the render method is expected to call `#to_json`. Funny for a while rails deprecated using `#to_json` in favor of `#as_json`. Times change I guess. Anyway, it is my understanding that `#as_json` should return a Hash, Array, or one of the primitive JSON types.

Okay back to you issue. First question is, "does it work with the default rails?". I think the `Hash.to_json` method does not call `#to_json` on it's members. Oj mimics that behavior. It might be different depending on whether the the `Hash.to_json` has been monkey patched. It gets a bit tricky knowing what patch is in effect.

How about an alternative approach. I haven't tried this myself but it seems like it should work. Either subclass `Oj::StringWriter` or make a class that contains it. Have the StringWriter build the complete JSON including the `key` in the example you gave. Alias the `#to_json` to call `#to_s`. Does that make sense to you?

**Comment by ElMassimo** [comment]
Thanks for your response! Makes sense, I was expecting to use these "wrapped values" in a backwards compatible way, so that I could use the string writer only in parts of the code where performance is critical.

With the current configuration, looks like `to_json` _is_ calling `to_json` on members, but it seems to be doing it recursively until it hits a primitive type. Tracing it I get:

```ruby
tracer.enable { { key: JsonValue.new('2') }.to_json }
[28, :c_call, Class, :new]
[36, :c_call, Kernel, :is_a?]
[22, :c_call, #<Class:Oj::Rails::Encoder>, :new]
[22, :c_call, Oj::Rails::Encoder, :encode]
"{\"key\":\"2\"}"
```

The ideal result would be to achieve a `{"key":2}` result instead. I will keep debugging, and post an update.

**Comment by ohler55** [comment]
Thanks. The more info you can give me the easier it will be to make changes if they are needed.

**Comment by ElMassimo** [comment]
After reading the [`rails.c`](https://github.com/ohler55/oj/blame/master/ext/oj/rails.c) source code, I have a better understanding of the internals, and the behavior that causes double encoding in the use case described above.

### Description :book:
When using `Oj.optimize_rails`, `Oj` will honor the `as_json` interface used by _ActiveSupport_’s default encoder, which will always call `as_json` on an object before serializing it to JSON.

`Oj` will detect if `as_json` returns the same object (to prevent a loop), in which case it will use the default strategy, which will call [`to_s`](https://github.com/ohler55/oj/blame/master/ext/oj/rails.c#L1415) on the object, _and JSON-encode_ the resulting String.

### Recycling an `Oj::StringWriter` string ♻️
There’s currently no way to reuse a previous JSON-encoded string when using `Oj::Rails::Encoder`.

From all the [existing strategies](https://github.com/ohler55/oj/blame/master/ext/oj/rails.c#L1419-L1446), the only one that does something similar is the [`BigDecimal` one](https://github.com/ohler55/oj/blame/master/ext/oj/rails.c#L557), which uses [`oj_dump_raw`](https://github.com/ohler55/oj/blame/master/ext/oj/rails.c#L218) (can’t subclass `BigDecimal` as a hack since Ruby doesn’t define `allocate` for that class).

### Feature Request :sparkles:
It would be very useful if the library defined a custom strategy for `Oj::StringWriter` objects, that invokes `to_s` on the writer object, and appends the returned JSON string in the buffer **without encoding it** by using `oj_dump_raw`.

That would make it easy to optimize a critical path using `Oj::StringWriter` instead of creating intermediate `Hash` and `Array` objects (which need to be allocated and garbage collected).

Probably not valid C, but some pseudocode with the idea:

```c
static void
dump_string_writer(VALUE obj, int depth, Out out, bool as_ok) {
  volatile VALUE rstr = rb_funcall(obj, oj_to_s_id, 0);
  const char *str = rb_string_value_ptr((VALUE*)&rstr);
  oj_dump_raw(str, (int)RSTRING_LEN(rstr), out);
}
```

### Alternative Proposal
Providing an `Oj::JsonValue` Ruby class that can be subclassed, which is encoded with the same strategy described above (`to_s` and `oj_dump_raw`), would provide maximum flexibility, and make it easier to create serialization libraries that leverage `Oj`’s excellent performance.

That class would allow to reuse `Oj::StringWriter` strings, as well as provide an escape hatch for objects that need to be encoded differently.

### Notes :pencil:
If adding `Oj::JsonValue` as a general purpose way to reuse existing JSON, it would then make sense to skip Ruby for `Oj::StringWriter` objects, and dump the buffer directly. Again, probably incorrect pseudocode:

```c
  StrWriter sw = (StrWriter)DATA_PTR(obj);
  oj_dump_raw(sw->out.buf, sw->out.cur - sw->out.buf, out);
```

**Comment by ohler55** [comment]
I was hesitant to break from rails behavior when in rails mode but I think there is a way to provide a `Oj::JsonValue` or `Oj::StringWriter` that will also work with rails with the same results as when using Oj. Maybe not as efficiently but at least the behavior would be the same. I'll have to experiment a bit but `Oj::StringWriter` had a `#to_json` method that might do it. If you want to try adding `#to_json` as an alias to `#to_s` and see what rails does with it I can certainly make Oj handling as you suggested in the note. If that doesn't work then maybe something like the `Oj::JsonValue` would be better as a special case that differs from rails behavior.

**Comment by ElMassimo** [comment]
Thank you for considering it!

Will investigate tomorrow, but I'm pretty sure Rails default encoder will always call `as_json` until it gets a primitive value, similar to what `Oj::Rails::Encoder` is doing. As a result, it will also double encode existing JSON (though it can be monkeypatched to achieve that, since it's written in Ruby). 

The current behavior for `Oj::StringWriter` objects when using `Oj::Rails::Encoder` always outputs `{}`, regardless of the contents in it's internal buffer; it's not likely that anyone is using it in practice. Changing the behavior would open up a lot of options for libraries such as Panko, and the one that I'm writing myself.

Adding `Oj::JsonValue` is even more flexible, allowing to reuse cached JSON, or JSON provided by an external API, without having to parse it first.

That said, we could keep the behavior opt-in, by having to manually call `Oj::Rails.optimize(Oj::StringWriter)` to use the suggested behavior, without introducing a breaking change for existing users.

**Comment by ohler55** [comment]
I agree an option is the way to go. I'll put something together later this week.

**Comment by ElMassimo** [comment]
Did some [initial work](https://github.com/ohler55/oj/compare/master...ElMassimo:master), by adding an `oj_string_writer_to_s` option, but still think it makes sense to create an `Oj::JsonValue`.

Let me know if you would prefer to implement it differently, at the moment I have only added support for the option in _Rails mode_.

**Comment by ohler55** [comment]
Looked at your initial work. I thought you wanted the StringWriter to provide not a quoted string but raw JSON. That would mean not using `#as_json` but an internal cheat for StringWriter and maybe others (JSONValue approach) so that there is some special method to call or maybe just the `#to_json`. I'm hesitant to use `#to_json` though as it has a lot of baggage with rails and the JSON gem that would cause unexpected behaviour with the monkey both do.

**Comment by ohler55** [comment]
I was thinking of `raw` option that allowed someone to specify which classes would be treated as a special case. Those classes would implement the `#raw_json` method that would take a depth and indent argument. The method would have to return a string. StringWriter would implement that but would also go under the covers and use the internal buffer directly for a performance boost.

I thought about just looking for a `#raw_json` method instead. That might be a better approach. What do you think?

**Comment by ElMassimo** [comment]
In the code above, `as_json` is only used when the `StringWriter` is a Hash or Array value. We need it because by default it would return an empty object `{}`, so it would never hit the custom dump function.

However, that change __is__ backwards compatible, no risk there.

Detecting a `raw_json` method seems like a reasonable strategy, but I'm not sure about the performance implications an extra `respond_to?` check would have. If it's acceptable, then it's definitely a flexible solution, and if `StringWriter` implements an optimized version, even better :smiley:

Using `raw_json` means we wouldn't need a special `JsonValue` object, since any class would be able to implement that protocol and achieve what we need (reuse existing JSON).

**Comment by ElMassimo** [comment]
If we want to skip the extra `respond_to?` check, then registering classes with a `raw` option like you mentioned first seems reasonable, not sure if the performance difference will justify the additional complexity though.

**Comment by ohler55** [comment]
If the `#respond_to` approach seems better the overhead is negligible.

**Comment by ohler55** [comment]
Please try the `raw_json` branch.

**Comment by ElMassimo** [comment]
Excellent, I will test it tomorrow, thanks!

**Comment by ElMassimo** [comment]
After some preliminary testing, it looks great!! Running the full test suite so check for incompatibilities, will post an update in an hour.

At the moment I'm ignoring the `depth` and `indentation` options (`def raw_json(*)`), since I'm not prettifying the output, but they might make sense for other use cases.

**Comment by ohler55** [comment]
The `depth` and `indent` will not change the StringWriter output but thought it best to provide them for classes that might.

**Comment by ElMassimo** [comment]
Found an incompatibility, which is the one mentioned above:

> `as_json` is only used when the `StringWriter` is inside a Hash or Array object. We need it because by default it would return an empty object `{}`, so it would never hit the custom dump function.

This happens when using the `rails` mode, since the contract is that for any `Hash` or `Array` value, `as_json` will be called before serializing the value.

I had [patched that](https://github.com/ohler55/oj/commit/1fe39ce6b9ab8d43ca7d741465926b88b0615d85) in my branch. Since the default behavior in Rails mode always renders `{}`, I think implementing `as_json` in `Oj::StringWriter` is a better default, regardless of whether `use_raw_json` is being used.

**Comment by ohler55** [comment]
The problem to work around there is that `as_json` expects an object that is a JSON primitive (Ruby Hash, Array, String, etc). If a string is returned it would be treated as a string and quoted which I don't think is what you want.

Some thoughts on that.

1) StringWriter does not and should not build a JSON object. It only builds a JSON string. When `as_json` is called it could parse that string to return a Ruby JSON primitive like Hash or Array. This is not at all efficient of course.

2) Have the StringWriter `#as_json` return itself. This is the most efficient. Not sure what that will do to any other parser though.

**Comment by ElMassimo** [comment]
Correct, #2 is the option I'm using with `Oj::Rails::Encoder`, because it will handle it gracefully when _any_ object returns `self`, not just primitive values.

It wouldn't work with the default encoder in _ActiveSupport_, which is why I suggest that we only define `as_json` for `StringWriter` if using `Oj.optimize_rails`. Else it might be puzzling for users to get a different result when using `Oj.dump(results: writer)` than when using `Oj.dump(writer)` if using the `rails` mode.

That said, I can just monkeypatch `def as_json; self` in `Oj::StringWriter`, which is working great with the `raw_json` branch :smiley:

**Comment by ohler55** [comment]
Fair enough. option 2 with `Oj.optimize_rails`. I'll do that tonight and then ask you for a final check before releasing.

**Comment by ElMassimo** [comment]
Sounds good, thank you Peter!

**Comment by ohler55** [comment]
Giving it more thought. I think this will work. If the containing Hash or Array is optimized or if the `as_json` options is turned off in combat more Oj will walk the Hash, Array, or Object attributes and will use the `raw_json` of StringWriter or other classes that implement `raw_json`. Otherwise, if the `as_json` is called it will do the nasty parse and return Ruby instances.

Its going to take a few test cases to verify that works.

**Comment by ElMassimo** [comment]
Cool, to provide some ideas for test cases, I'm using:
```ruby
Oj.optimize_rails
Oj.default_options = { mode: :rails, use_raw_json: true }
```
and without monkeypatching `as_json` for `Oj::StringWriter` to return `self`, [this code path](https://github.com/rails/rails/blob/b9ca94caea2ca6a6cc09abaffaad67b447134079/activesupport/lib/active_support/core_ext/object/json.rb#L56) is used for any `Oj::StringWriter` instances that are `Hash` or `Array` values.

**Comment by ohler55** [comment]
I see, so if starting from a call to `as_json` on an Object Oj doesn't have a chance to get in the middle.

Do you know where the rails code is that handles an object like StringWriter returning itself?

**Comment by ElMassimo** [comment]
As mentioned above, the default encoder [doesn't handle non-primitive values](https://github.com/rails/rails/blob/b9ca94caea2ca6a6cc09abaffaad67b447134079/activesupport/lib/active_support/json/encoding.rb#L72-L98), but Oj [does](https://github.com/ohler55/oj/blob/develop/ext/oj/rails.c#L530-L532).

**Comment by ohler55** [comment]
I misunderstood, I thought you said rails handed it. Ok, must switch on optimized then. Thanks for the patience.

**Comment by ElMassimo** [comment]
No problem, I see that behavior as a shortcoming in ActiveSupport's encoder, it [duplicates the code](https://github.com/rails/rails/blob/b9ca94caea2ca6a6cc09abaffaad67b447134079/activesupport/lib/active_support/json/encoding.rb#L89-L90), instead of checking if `as_json` returns the same object, which would be both faster and more flexible.

**Comment by ohler55** [comment]
Pushed a version that I think does what we want. Please give it a run.

**Comment by ElMassimo** [comment]
Run some quick tests, worked like a charm! 😃 

Will post an update in an hour after running the entire test suite.

Thank you again for your patience, and your hard work on this excellent library!

**Comment by ohler55** [comment]
Great. If all is good I'll release tomorrow.

**Comment by ElMassimo** [comment]
All green! Thank you so much for making this change!

**Comment by ohler55** [comment]
Released v3.8.0.

**Comment by ElMassimo** [comment]
Having `StringWriter` play along with `Oj::Rails::Encoder` is super useful, thanks again!

Closed in #547, [see the changelog for details](https://github.com/ohler55/oj/blob/develop/CHANGELOG.md#380---2019-07-17).

**Comment by ElMassimo** [comment]
@ohler55 I have open sourced a library that leverages the `raw_json` option introduced in #547, I've named it [`oj_serializers`](https://github.com/ElMassimo/oj_serializers), and it's meant as an Active Model Serializers alternative.

Thanks again for your work on the library!
